Lib/pty.py

```diff
@@ -170,3 +175,128 @@ def spawn(argv, master_read=_read, stdin_read=_read):
 
     os.close(master_fd)
     return os.waitpid(pid, 0)[1]
+
+# Author: Soumendra Ganguly.
+SIGWINCH = signal.SIGWINCH
+
+def _login_pty(master_fd, slave_fd):
+    """Given a pty, makes the calling process a session leader, makes the pty slave
+    its controlling terminal, stdin, stdout, and stderr. Closes both pty ends."""
+    # Establish a new session.
+    os.setsid()
+    os.close(master_fd)
+
+    try:
+        fcntl.ioctl(slave_fd, termios.TIOCSCTTY) # Make the pty slave the controlling terminal.
+    except:
+        os.close(slave_fd)
+        raise
+
+    # Slave becomes stdin/stdout/stderr.
+    os.dup2(slave_fd, STDIN_FILENO)
+    os.dup2(slave_fd, STDOUT_FILENO)
+    os.dup2(slave_fd, STDERR_FILENO)
+    if (slave_fd > STDERR_FILENO):
+        os.close(slave_fd)
+
```

Comment:
C code equivalent to "_winresz" is

struct winsize w;
if (ioctl(STDIN_FILENO, TIOCGWINSZ, &w) == 0)
                ioctl(pty_slave, TIOCSWINSZ, &w);

Notice that the above code checks if the TIOCGWINSZ call is a success before
making the TIOCSWINSZ call. Therefore, "_winresz" must always be called
in a try block. References:

1. https://www.man7.org/linux/man-pages/man4/tty_ioctl.4.html